gl_engine/shader: Introduced binary shader caching

wenjieshen opened the pull request:
We've implemented a comprehensive binary shader caching system to significantly improve OpenGL rendering performance on macOS by eliminating redundant shader compilation overhead on subsequent application runs.

This feature introduces the `GlShaderCache` class that provides persistent storage and retrieval of compiled OpenGL shader binaries. The system automatically caches compiled shader programs to the user's cache directory using a hash-based naming scheme and attempts to load cached binaries before falling back to source compilation.

Key implementation details:
- Utilizes OpenGL binary shader format support via `glGetProgramBinary`/`glProgramBinary` extensions
- Stores cached binaries in the System cache folder on macOS and Windows; other platforms are testing
- Integrates with the existing `GlProgram` compilation pipeline

The caching mechanism works transparently:
1. When a shader program is first compiled, the binary is automatically cached
2. On subsequent runs, the system attempts to load the cached binary
3. If successful, compilation time is virtually eliminated
4. If cache miss or load failure occurs, falls back to normal source compilation

Performance benefits:
- Eliminates shader compilation stutters during runtime
- No API changes required - caching works automatically

issue: https://github.com/thorvg/thorvg/issues/2884

hermet commented:
> Binary shader support is an OpenGL 4.1+ feature, while our requirement is OpenGL 3. I added a helper function glIsProgramBinarySupportAvailable() in tvgGl.h to check runtime availability. This is the first runtime capability check in the project, so I’m not sure if this approach aligns with our design principles.

@huming2207 Looks correct, if the binary shaders doesn't working, we can fallback to a normal scenario.

huming2207 commented:
> @huming2207 Could you please double-check this as well? a few systems might use this GL_ARB_get_program_binary for gl3 case.

@hermet  Sorry I'm not quite familiar with OpenGL, I'm a firmware guy and I know more on low-level stuff only 😅

hermet commented:
> Understood, I’ll pause the PR.

@wenjieshen I mean we can hold support v3 for binary shaders. I didn't mean hold this PR. Please go proceed. :)  Thanks.

hermet commented on src/renderer/gl_engine/tvgGl.cpp:
```diff
@@ -465,6 +465,7 @@ PFNGLGENVERTEXARRAYSPROC                     glGenVertexArrays;
 //GL_VERSION_3_1
 PFNGLGETUNIFORMBLOCKINDEXPROC      glGetUniformBlockIndex;
 PFNGLUNIFORMBLOCKBINDINGPROC       glUniformBlockBinding;
+
```
please don't change code without any specific purpose.

hermet reviewed (commented):
@wenjieshen Please check comments and please respect thorvg coding convention.
Thanks.

hermet commented on src/renderer/gl_engine/tvgGl.cpp:
```diff
@@ -820,6 +830,22 @@ bool glInit()
+        glProgramBinarySupportAvailable = true;
+    }
+#else
+    // OpenGL 4.1+ has these functions in core
```
please remove helpfuless comments

hermet commented on src/renderer/gl_engine/tvgGl.cpp:
```diff
@@ -835,4 +861,9 @@ bool glTerm()
     return true;
 }
 
+bool glIsProgramBinarySupportAvailable()
```
`bool glProgramBinarySupport()`

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,260 @@
+
+uint32_t GlShaderCache::read(const char* vertSrc, const char* fragSrc)
+{
+#if defined(THORVG_FILE_IO_SUPPORT) && !defined(__EMSCRIPTEN__)
```
--unused parameters warnings
Suggestion:
```
    if (!vertSrc || !fragSrc) return 0;

#if defined(THORVG_FILE_IO_SUPPORT) && !defined(__EMSCRIPTEN__)
```

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,260 @@
+    auto cachePath = path(vertSrc, fragSrc);
+    if (!cachePath) return 0;
+
+    // Check if cache file exists
```
--unhelpful comments

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,260 @@
+        return 0;
+    }
+
+    // Create program object
```
--unhelpful comments

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,260 @@
+        return 0;
+    }
+
+    // Read binary data
```
--unhelpful comments

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,260 @@
+        return false;
+    }
+
+    // Write format and length first
```
-- unnessary comment

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,260 @@
+        return false;
+    }
+
+    // Get program binary length
```
-- unnessary comment

hermet reviewed (commented):
Most your comments are unnecesary. Don't need to add comments for saying code itself.
We'd rather leave comments when it needs a unusual, special context and purposes.

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.h:
```diff
@@ -0,0 +1,38 @@
+    static uint32_t read(const char* vertSrc, const char* fragSrc);
+
+    static const char* path(const char* vertSrc, const char* fragSrc);
+    static uint32_t hashString(const char* str);
```
-> hash()

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,244 @@
+        return false;
+    }
+
+    FILE* file = fopen(cachePath, "wb");
```
auto

hermet commented on src/renderer/gl_engine/tvgGl.cpp:
```diff
@@ -809,6 +815,10 @@ bool glInit()
     // GL_FUNCTION_FETCH(glGetActiveUniformBlockName, PFNGLGETACTIVEUNIFORMBLOCKNAMEPROC);
     GL_FUNCTION_FETCH(glUniformBlockBinding, PFNGLUNIFORMBLOCKBINDINGPROC);
 
+    // GL_ARB_get_program_binary (optional, may not be available on all platforms)
```
//TODO: GL_ARB_get_program_binary  ...

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,244 @@
+
+bool GlShaderCache::write(uint32_t programId, const char* vertSrc, const char* fragSrc)
+{
+    if (!programId || !vertSrc || !fragSrc) return false;
```
--Unneccesary conditions
 (!vertSrc || !fragSrc)

hermet commented on src/renderer/gl_engine/tvgGlProgram.cpp:
```diff
@@ -35,10 +36,21 @@ uint32_t GlProgram::mCurrentProgram = 0;
+    uint32_t progObj = 0;
+
+#ifndef __EMSCRIPTEN__
+    progObj = GlShaderCache::read(vertSrc, fragSrc);
```
Simpler code
```
mProgObj = GlShaderCache::read(vertSrc, fragSrc);
if (mProgObj != 0) return;
```

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,244 @@
+#endif
+}
+
+bool GlShaderCache::write(uint32_t programId, const char* vertSrc, const char* fragSrc)
```
-> void. 
return value is not used anywhere.

wenjieshen commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,283 @@
+
+static char cachePath[PATH_MAX];
+
+uint32_t GlShaderCache::hashString(const char *str)
```
Thank you. I am setting up environment to ensure if it works in this case.

wenjieshen commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,244 @@
+    const char* home = getenv("HOME");
+    if (!home) return nullptr;
+
+    #ifdef _WIN32
```
Thanks, I’ll fix the indentation. One clarification: this block is already inside `#if defined(THORVG_FILE_IO_SUPPORT) && !defined(__EMSCRIPTEN__)`. Should we indent mixed nested preprocessor directives too, or only the full macro code snippet per our style?

wenjieshen replied:
Got it. This issue will be addressed when the Windows support is merged into the upcoming version.

SergeyLebedkin reviewed (commented):
see comments

wenjieshen commented:
# Shader Binary‐Cache Experiment Results

## Overview

This analysis examines shader creation and caching performance across 30 repeated executions, each compiling 16 shaders.  
Four distinct timing datasets were analyzed:

- **NVIDIA Binary Shader Cache (available):** Shader creation when the driver’s built-in binary shader cache is active.  
- **Without Binary Shader Cache:** Shader creation times when the driver cache is unavailable or disabled.  
- **Cache read:** Loading precompiled shaders from a custom in-house cache system.  
- **Cache write:** Writing shader binaries to that same custom cache.

## About the NVIDIA Auto Binary Shader Cache

The **NVIDIA Auto Binary Shader Cache** stores compiled shader binaries on disk so they can be reused across application runs instead of being recompiled from source each time.  
This cache is **automatically managed by NVIDIA drivers** and is **available by default** on most driver versions. However, the exact behavior and availability may differ depending on the **GPU vendor** and **driver version**.  

When present, the driver transparently saves and restores shader binaries from its cache directory, substantially reducing load times for subsequent runs.

> **Reference:** [NVIDIA DRIVE OS 5.2 Linux SDK Development Guide – Binary Shaders](https://docs.nvidia.com/drive/drive-os-5.2.0.0L/drive-os/DRIVE_OS_Linux_SDK_Development_Guide/Graphics/graphics_binary_shaders.html#wwpID0EOHA)

## Summary Statistics

| Metric | Binary Shader Cache | No Cache | Cache Read | Cache Write |
| --- | --- | --- | --- | --- |
| Mean (µs) | **585** | **10,952** | 198 | 102 |
| Median (µs) | 116 | 4,454 | 126 | 58 |
| 90th Percentile (µs) | 475 | 41,797 | 484 | 110 |
| 95th Percentile (µs) | 788 | 51,207 | 728 | 137 |
| 99th Percentile (µs) | 5,868 | 51,978 | 929 | 980 |
| Maximum (µs) | 50,400 | 57,287 | 1,102 | 1,383 |
| Minimum (µs) | 87 | 3,159 | 83 | 35 |

## Comparative Results

### 1. Shader Compilation Time
- The **Binary Shader Cache** reduces mean shader compilation time from roughly **10.95 ms to 0.59 ms**, achieving an average **54× speed-up** across runs.
- The median drop is even more dramatic, from 4.45 ms to 0.12 ms — roughly **37× faster**.

### 2. Cache Read vs Write Performance
- Reading shaders from the in-house cache averages **198 µs**, about **3× faster** than writing them (102 µs).
 **Note:** The write process is performed **only during the first run**, as it stores new shader binaries for reuse. Subsequent executions only perform read operations.

### 3. Distribution & Variability
- Without the Binary Shader Cache, shader creation exhibits long-tail latency, with outliers exceeding **50 ms**.
- With caching available, almost all shader compilations finish under **1 ms**, except during first-time runs.
- In-house cache read/write times remain consistently under **200 µs**.

---

## Visualizations

**Average per-run shader creation time comparison**  
<img width="1000" height="600" alt="per_run_average_comparison" src="https://github.com/user-attachments/assets/29ccba0b-f4e8-499f-98c6-6b0b775ee3df" />

**Distribution of all operation times**  
<img width="1000" height="600" alt="per_shader_average_comparison" src="https://github.com/user-attachments/assets/91afddcc-901d-466b-af5b-8fae2bbe5e2d" />

**Speed-up ratio (no-cache ÷ cache)**  
<img width="800" height="600" alt="distribution_boxplot" src="https://github.com/user-attachments/assets/ffd0fde5-3aeb-45b5-98af-ad38a6a0f712" />

hermet reviewed (commented):
@wenjieshen Thanks, Please check more nit-picks.  
+ Please note that ThorVG don't handles memory alloc fails.

wenjieshen commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,271 @@
+    if (!cachePath) return;
+
+    GLint numFormats = 0;
+    glGetIntegerv(GL_NUM_PROGRAM_BINARY_FORMATS, &numFormats);
```
It works well. Thanks.

hermet commented on src/common/tvgEnv.h:
```diff
@@ -0,0 +1,122 @@
+        } else {
+            auto home = getenv("HOME");
+            if (!home) return false;
+            snprintf(cacheDir, maxLen, "%s/.cache/thorvg/shaders", home);
```
This function needs to have `%s/.cache/thorvg/`

2 options, we can have:
1. you can just put the shader files in thorvg without `shaders`
2. you can append the `shaders` subfolder from gl engine

I bet on no.1

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,175 @@
+#endif
+
+#ifndef __EMSCRIPTEN__
+static char cachePath[PATH_MAX];
```
Let's remove this static var.

hermet commented on src/common/tvgEnv.h:
```diff
@@ -0,0 +1,119 @@
+
+namespace tvg
+{
+    inline bool cacheDir(char* cacheDir, size_t maxLen)
```
cacheDir -> cachedir()

hermet commented on src/renderer/gl_engine/tvgGlShaderCache.h:
```diff
@@ -0,0 +1,36 @@
+#include "tvgGlCommon.h"
+
+struct GlShaderCache
+{
```
`static bool support;`

SergeyLebedkin reviewed (approved):
✅windows 11: OK
✅ubuntu (wsl2): OK
✅ubuntu (h/w): OK
✅macos (intel): OK (there are no supported binary formats)

hermet commented:
The file IO is expensive relatively. Current shaders are more than 10, it needs a consideration before pushing this. That would be the best if we can introduce the one single binary form....

hermet commented:
**1. Verifying Shaders**
- Current: Use a Hash & ThorVG version
- After: Use the ThorVG version only. ie) `CACHE_FOLDER/thorvg/gl_shaders_VERSION_INFO.bin`
- Later we can introduce a binary shader feature toggle option. 
- - On dev, we can disable the binary shader, but we can enable it manually when necessary.
- - Enable the binary shader feature with the official release binary.

**2. Lazy Loading or Immediate Loading**
 - Lazy Loading needs to hold the binary shader memory, We need to confirm the total size of the memory first.

wenjieshen commented:
> **1. Verifying Shaders**
> 
> * Current: Use a Hash & ThorVG version
> * After: Use the ThorVG version only. ie) `CACHE_FOLDER/thorvg/gl_shaders_VERSION_INFO.bin`
> * Later we can introduce a binary shader feature toggle option.
> * * On dev, we can disable the binary shader, but we can enable it manually when necessary.
> * * Enable the binary shader feature with the official release binary.
> 
> **2. Lazy Loading or Immediate Loading**
> 
> * Lazy Loading needs to hold the binary shader memory, We need to confirm the total size of the memory first.

@hermet 
I understand. Thanks for the suggestion.  
We can roll back this version if the loading efficiency gain doesn’t justify the additional memory overhead. On average, loading a single discrete shader takes 0.59 ms, while loading the combined binary file (containing all shaders) takes 0.87 ms. Since each example may require between 17 and 89 shaders, the unified loading approach can improve total loading performance by roughly **11× to 60×**, depending on how many shaders are needed.
### Memory Usage  
After combining the binary shader files, the **maximum additional memory usage** observed across all examples is **857.21 KB**.

### Shader Loading Time (per example)  
- **Mean (Average):** 0.867 ms  
- **Median (50th Percentile):** 0.877 ms  
- **Standard Deviation:** 0.175 ms  
- **Variance:** 0.031  
- **Minimum:** 0.183 ms (Fastest load time — *Animation.exe*)  
- **Maximum:** 1.322 ms (Slowest load time — *StrokeLine.exe*)

### Accumulated Write Time(per example)
- Blending.exe: $10.897$ ms
- Accessor.exe: $4.334$ ms
- SceneEffects.exe: $1.415$ ms
- Duplicate.exe: $0.455$ ms
- EffectDropShadow.exe: $0.343$ ms
### Shader Writing Time (per shader)  
- **Mean (Average):** 0.203 ms  
- **Median (50th Percentile):** 0.174 ms  
- **Standard Deviation:** 0.090 ms  
- **Variance:** 0.008  
- **Minimum:** 0.126 ms (Fastest write time)  
- **Maximum:** 0.670 ms (Slowest write time)

wenjieshen commented:
@hermet Hi Hermet, there is a data of more practical testing
Exec | #Shaders | Original CSV total (ms) | Binary CSV total (ms) | Speedup (orig / binary)
-- | -- | -- | -- | --
0 | 80 | 588.421 | 13.490 | 43.62×
1 | 80 | 571.231 | 12.115 | 47.15×
2 | 80 | 571.627 | 10.817 | 52.85×
3 | 80 | 628.088 | 10.865 | 57.81×
4 | 80 | 579.044 | 11.483 | 50.43×
5 | 80 | 577.148 | 10.702 | 53.93×
6 | 80 | 582.733 | 10.572 | 55.12×
7 | 80 | 582.633 | 11.139 | 52.31×
8 | 80 | 578.692 | 11.152 | 51.89×
9 | 80 | 579.014 | 11.394 | 50.82×
10 | 80 | 550.348 | 9.966 | 55.22×

### Statistical comparison (using the 10 complete executions, 80 shaders each)

- Original (glprogram_timing.csv):

1. Mean total per execution: 583.9 ms

2. Std dev: 16.4 ms

3. Min / Max: 571.2 ms / 628.1 ms

- Binary (glprogram_timing_binary.csv):

1. Mean total per execution: 11.37 ms

2. Std dev: 0.87 ms

3. Min / Max: 10.57 ms / 13.49 ms

Relative:

Average speedup over the first 10 executions: ≈ 51.6× faster using the binary timings.

On average you save about ~572 ms per execution (583.9 ms − 11.37 ms).
[glprogram_timing.csv](https://github.com/user-attachments/files/23539988/glprogram_timing.csv)
[glprogram_timing_binary.csv](https://github.com/user-attachments/files/23539990/glprogram_timing_binary.csv)

wenjieshen commented:
On the Windows platform, the nVidia driver automatically caches binary shaders. Disabling this setting provides more accurate values.
<img width="1622" height="1177" alt="image" src="https://github.com/user-attachments/assets/1f247af7-2d74-493d-86f7-4ac09dc75b72" />

wenjieshen committed:
gl_engine: Add binary shader caching support

Implement binary shader caching system to improve OpenGL performance by:
- Add GlShaderCache class with read/write functionality for compiled shaders
- Support OpenGL binary shader format
- Add proper OpenGL ARB extension loading for glGetProgramBinary/glProgramBinary
- Integrate caching into GlProgram compilation pipeline

This reduces shader compilation overhead on subsequent application runs,
significantly improving startup performance for OpenGL-based rendering.

Related issue: #2884

hermet commented on src/common/tvgEnv.h:
```diff
@@ -0,0 +1,119 @@
+
+#ifdef _WIN32
+    #ifndef WIN32_LEAN_AND_MEAN
+        #define WIN32_LEAN_AND_MEAN
```
our config has this by default:

<img width="743" height="102" alt="image" src="https://github.com/user-attachments/assets/adc6a062-aa0c-4302-b12b-182de5182e4a" />

no need to add define.

Here portable if-def can be moved in `tvgCommon.h`

hermet reviewed (commented):
@wenjieshen please check a comment.

hermet commented:
At boot time: 3%+ 
Only Shader loading part: 12%+

(Tested on Intel Arc Graphics)

@wenjieshen Unlike frame-time enhancements, a 3% improvement in boot time is somewhat underwhelming, especially when considering the maintenance overhead. Let's put this optimization on hold for now. We may see better performance improvements by considering a single, unified binary shader. However, since the overall burden is under 1ms, the improvement might not be very noticeable. Thanks for your trial & effort.

wenjieshen commented on src/renderer/gpu_engine/gl/tvgGlShaderCache.cpp:
```diff
@@ -0,0 +1,293 @@
+        cached.data = binaryData;
+        cached.length = entryHeader.length;
+        cached.format = entryHeader.binaryFormat;
+        cache[entryHeader.renderType] = cached;  // The last entry will overwrite the previous one with the same renderType
```
It is fair. I updated.

hermet commented:
@wenjieshen Could you please rebase this one more time? Thanks.

Nor-s commented:
I tested this using **NVIDIA's cache-disable flag**. 

I ran it on the **Shapes**, **Blending**, and **SceneEffects** examples, 
and measured performance by inserting timing code at `GlRenderer::initShaders()`, `draw`, and `sync`.

I tested it like this:
```
__GL_SHADER_DISK_CACHE=0 ./builddir/src/SceneEffects gl
```

> [!IMPORTANT]
> If i enable the **NVIDIA cache** on **main**, the numbers come out **almost identical to the Binary Shader fetch**.


> [!NOTE]
> [test branch(cherry-pick binary shader feature)](https://github.com/thorvg/thorvg/tree/nor-s/binary-shader-test)
> **ENV: Arch linux, AMD 7600, Nvidia 3060ti**
> **Averaging method:**
> - **Main:** average of all 9 runs.
> - **Binary Shader:** average of the 6 **cache-hit** runs only. The first run of each iteration is excluded, since that run compiles the shaders and bakes the cache (including it would dilute the cache effect). 
> - And I cleared the cache for each individual test run.




### Shapes

| Stage | Change |
|---|---|
| init_shaders |  **−98.8%** |
| first canvas draw |  **−98.7%** |
| first canvas sync  | **+65.6%** |
| Boot (ms)  | **−53.2%** |

### Blending

| Stage | Change |
|---|---|
| init_shaders | **−98.2%** |
| first canvas draw |  **−85.6%** |
| first canvas sync | **+276.8%** |
| Boot (ms) |  **−44.6%** |


### Scene Effects

| Stage |  Change |
|---|---|
| init_shaders |  **−98.6%** |
| first canvas draw  | **−86.8%** |
| first canvas sync  | **+19.9%** |
| Boot (ms)| **−49.4%** |

hermet commented:
@wenjieshen please rebase it.
